### ISSUE TITLE
ISPC uses scalar SSE instructions to convert floats to doubles when compiling for AVX, AVX2 and AVX512

### ISSUE THREAD

**Issue by pbrubaker** [ispc/ispc#2147]
The title pretty much says it all.  ISPC is not doing wide conversions from float to double.  Padding doesn't seem to affect it either.

https://ispc.godbolt.org/z/Th6G5xxbn

**Comment by JeffRous** [comment]
Fake varying can be abused to do this?

https://ispc.godbolt.org/z/j6cW8nfeh

**Comment by pbrubaker** [comment]
Yeah, that does collapse the conversion instructions, but ideally I'd like to see a `vmovmaskps` instead of the inserts.

**Comment by pbrubaker** [comment]
Hey @dbabokin is there any chance we can look at this one for the next release?  Some of the code I'm working on right now is relatively short and does promotion conversions from float to double often and I'm pretty sure this is affecting performance.

**Comment by pbrubaker** [comment]
I ran into this one again with integers today.

```
    uniform float reduce_weight = reduce_add(weight);
    
    uniform float <4> reduce_col;
    reduce_col.x = reduce_add( col.x );
    reduce_col.y = reduce_add( col.y );
    reduce_col.z = reduce_add( col.z );
    reduce_col.w = reduce_add( col.w );

    reduce_col = (reduce_col / reduce_weight) * 255.0f;

    uniform uint32<4> ret = (uint32<4>) reduce_col;
    return ret;  //return value here is uint8<4>
```

```
        vdivps  xmm0, xmm0, xmm1
        vmulps  xmm0, xmm0, dword ptr [rip + .LCPI2_5]{1to4}
        vcvttss2si      eax, xmm0
        vmovshdup       xmm1, xmm0              # xmm1 = xmm0[1,1,3,3]
        vcvttss2si      ecx, xmm1
        vpermilpd       xmm1, xmm0, 1           # xmm1 = xmm0[1,0]
        vcvttss2si      edx, xmm1
        vpermilps       xmm0, xmm0, 255         # xmm0 = xmm0[3,3,3,3]
        vcvttss2si      esi, xmm0
        vmovd   xmm0, eax
        vpinsrb xmm0, xmm0, ecx, 1
        vpinsrb xmm0, xmm0, edx, 2
        vpinsrb xmm0, xmm0, esi, 3
```
I'm not sure why the compiler isn't preferring ` _mm256_cvttps_epi32` or in this case `_mm_cvttps_epi32`.  Since we're truncating to 8-bits, those inserts could go away and we could just mask off the higher order bits.

This case is relatively expensive because in this kernel I'm doing this for every pixel in an image.

**Comment by pbrubaker** [comment]
@JeffRous I was dinking around with this again tonight and I was able to get the same output as you with much less casting.

https://ispc.godbolt.org/z/G3PYYrPos

```
uniform QVec3 SetQVec3(uniform const FVector4 fp0)
{
    varying float F = *((varying float* uniform)&fp0);

    const varying double D = F;  
    const uniform QVec3 Result = *((uniform double<4>* uniform)&D);

    return Result;
}
```

vs.

```
uniform QVec3 SetQVec3(uniform const FVector4 fp0)
{
	varying float B;
	*((varying float *uniform)&B) = *((varying float *uniform)&fp0);
	const varying double C = (double)B;
	uniform QVec3 Result;
	*((uniform QVec3 *uniform)&Result) = *((uniform QVec3 *uniform)&C);
	return Result;
}
```

I think it's quite a bit easier to read too as you don't have to do any casting on the left hand side.

**Comment by dbabokin** [comment]
From compiler engineering prospective the request sounds like "use vector instructions for short vector initialization". I.e.
The following code should imply vector semantic "by definition" (though it still has multiple questions):
```
const uniform float<4> f = {1.f, 2.f, 3.f, 4.f}; 
const uniform double<4> d = {f[0], f[1], f[2], 0.0};
```

Note that it's a bit different question than code gene for @pbrubaker code above, where the short vector is not **initialized**, but every individual component of the short vector is assigned instead:
```
    uniform float <4> reduce_col;
    reduce_col.x = reduce_add( col.x );
    reduce_col.y = reduce_add( col.y );
    reduce_col.z = reduce_add( col.z );
    reduce_col.w = reduce_add( col.w );
```

I two possible way to address this:
1. Redesign how code generation (note here this refers AST->LLVM IR code gen) is done for short vectors.
2. Do "optimization" that does vectorization of such case (possibly enabling LLVM SLP Vectorizer Pass is a solution).

I think solving (1), or at least clearly defining the behavior that we have for short vectors should be first step here.

**Comment by aneshlya** [comment]
Fixed by PR #2501 and PR #2531.
